### PR TITLE
accelerate mixed precision integrate

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Shift Trainer's mixed precision handling to accelerate. Having smaller PR for this instead of mixing this with DDP, FSDP and DeepSpeed changes.
2. Sharded DDP and Apex are cases not supported in Accelerate and because of this I'm unable to simplify further and delete chunks of code in Trainer.
3. No user-facing changes. User can now use `accelerate launch` for launching mixed precision training in Trainer. Example given below:
```
accelerate launch --mixed_precision="bf16" run_glue.py   --model_name_or_path bert-base-cased   --task_name $TASK_NAME   --do_train   --do_eval   --max_seq_length 128   --per_device_train_batch_size 32   --learning_rate 2e-5   --num_train_epochs 3   --output_dir ~/temp/$TASK_NAME/ --overwrite_output_dir
```

the previous usage via `python` or `torchrun` is same. 
```
python run_glue.py   --model_name_or_path bert-base-cased   --task_name $TASK_NAME   --do_train   --do_eval   --max_seq_length 128   --per_device_train_batch_size 32   --learning_rate 2e-5   --num_train_epochs 3   --output_dir ~/temp/$TASK_NAME/ --fp16 --overwrite_output_dir
```

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for working on this!
Re-Sharded DDP -> I think this library is dead and the new maintained way is FSDP directly in PyTorch so we can deprecate this and remove it in 2-3 minor versions (can be done in a separate PR)
Apex -> We will need to keep supporting it as a separate code path (or add support in Accelerate but not urgent).

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1760,6 +1768,13 @@ def _inner_training_loop(
         if delay_optimizer_creation:
             self.create_optimizer_and_scheduler(num_training_steps=max_steps)
 
+        # prepare using `accelerator` prepare
+        model, self.optimizer, self.lr_scheduler = self.accelerator.prepare(
+            self.model, self.optimizer, self.lr_scheduler
+        )
+        self.accelerator.print(f"{model=}\n{self.optimizer=}\n{self.lr_scheduler=}")
+        self.accelerator.print(f"{model.forward=}")
```
Comment (muellerzr) [inline src/transformers/trainer.py:1776]: Are all these `print` debugging statements or meant to be logged? If they're meant to be logged, let's use the `logger` and put it under `debug`

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -337,6 +339,11 @@ def __init__(
         self.deepspeed = None
         self.is_in_train = False
 
+        # create accelerator object
+        self.accelerator = Accelerator()
+        self.accelerator.print(f"{self.accelerator=}")
+        self.accelerator.print(f"{self.accelerator.state=}")
```
Comment (muellerzr) [inline src/transformers/trainer.py:345]: Same comment as below

**Review by muellerzr — APPROVED** [review]
LG2M! A few notes on the `print`'s

**Pull request merged**